module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+		}
+
+		logger.Debug().Msgf("queue has %d requests", e.fetchQueue.Len())
+
+		highestSealed = request.height
+		e.fetchNotifier.Notify()
+	}
+
+	e.notificationState.Sealed(highestSealed)
+	return nil
+}
+
+func (e *executionDataRequesterImpl) missingBefore(current uint64, end uint64) ([]fetchRequest, error) {
+	missedHeights := []fetchRequest{}
+
+	for current < end {
+		e.log.Debug().Msgf("adding missed sealed height: %d, latest: %d", current, end)
+		missedBlock, err := e.blocks.ByHeight(current)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get block for height %d: %w", current, err)
+		}
+
+		missedHeights = append(missedHeights, fetchRequest{
+			blockID: missedBlock.ID(),
+			height:  current,
+		})
+
+		current++
+	}
+
+	return missedHeights, nil
+}
+
+// Fetch Worker Methods
+
+func (e *executionDataRequesterImpl) fetchRequestProcessingLoop(ctx irrecoverable.SignalerContext) {
+
+	e.consumerMu.Lock()
+	workerID := e.workerCount
+	e.workerCount++
+	e.consumerMu.Unlock()
+
+	logger := e.log.With().Int("worker_id", workerID).Logger()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-e.stop:
+			return
+		case <-e.fetchNotifier.Channel():
+			logger.Debug().Msgf("fetching available requests")
+			e.processAvailableFetchRequests(ctx)
+			logger.Debug().Msgf("done fetching requests")
+		}
+	}
+}
+
+func (e *executionDataRequesterImpl) processAvailableFetchRequests(ctx irrecoverable.SignalerContext) {
+
+	e.log.Debug().Msg("processing fetch requests")
+	for {
+		// retries get priority since the seal height is most likely lower
+		item, ok := e.fetchRetryQueue.Pop()
+		if ok {
+			e.processFetchRequest(ctx, item.(fetchRequest))
+			continue
+		}
+
+		// Check for missing backfills
+
+		item, ok = e.fetchQueue.Pop()
+		if ok {
+			e.processFetchRequest(ctx, item.(fetchRequest))
+			continue
+		}
+
+		// return when queue is empty
+		return
+	}
+}
+
+func (e *executionDataRequesterImpl) processFetchRequest(ctx irrecoverable.SignalerContext, request fetchRequest) error {
+	logger := e.log.With().Str("block_id", request.blockID.String()).Logger()
+
+	logger.Debug().Msgf("processing fetch request for block %d", request.height)
+
+	executionData, err := e.byBlockID(ctx, request.blockID)
+
+	logger.Debug().Msgf("done processing fetch request for block %d", request.height)
+
+	if errors.Is(err, &state_synchronization.MalformedDataError{}) || errors.Is(err, &state_synchronization.BlobSizeLimitExceededError{}) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data id is valid, and not sign the receipt
+		// Crashing isn't a good idea since that would take down access to the network, but this should probably halt the network somehow
+		// maybe we can flag it as bad and not reattempt
+
+		// TODO: add metric
+		logger.Error().Err(err).Msg("HALTING REQUESTER: invalid execution data found")
+		close(e.stop)
```

Comment:
and don't forget to return here

Reply:
otherwise you will continue to the next block.

Reply:
Also, I actually do think from the perspective of the ExecutionDataRequester, this is a fatal error, so I think we should `ctx.Throw` a sentinel error. Then, the handler of this error can check this and not crash the node if it is this specific error.